hydra-cluster/src/Hydra/Cluster/Faucet.hs

```diff
@@ -56,16 +56,38 @@ seedFromFaucet ::
   IO UTxO
 seedFromFaucet node@RunningNode{networkId, nodeSocket} receivingVerificationKey lovelace tracer = do
   (faucetVk, faucetSk) <- keysFor Faucet
-  retryOnExceptions tracer $ submitSeedTx faucetVk faucetSk
-  waitForPayment networkId nodeSocket lovelace receivingAddress
+  payments <- findPaymentsWithSameLovelace
+  case payments of
+    Just currentPayments -> do
+      signedTx <- retryOnExceptions tracer $ submitSeedTx faucetVk faucetSk
+      void $ awaitTransaction networkId nodeSocket signedTx
+      newPayments <-
+        waitForPayments networkId nodeSocket lovelace receivingAddress
+      pure
+        $ UTxO.fromPairs
+          . filter
+            ( \(txIn, _) ->
+                not $ txIn `member` UTxO.inputSet currentPayments
+            )
+        $ UTxO.pairs newPayments
+    Nothing -> do
+      void $ retryOnExceptions tracer $ submitSeedTx faucetVk faucetSk
+      waitForPayments networkId nodeSocket lovelace receivingAddress
  where
+  findPaymentsWithSameLovelace =
+    failAfter 1 $
+      (Just <$> waitForPayments networkId nodeSocket lovelace receivingAddress)
+        `catch` \(_ :: SomeException) -> pure Nothing
+
   submitSeedTx faucetVk faucetSk = do
     faucetUTxO <- findFaucetUTxO node lovelace
     let changeAddress = ShelleyAddressInEra (buildAddress faucetVk networkId)
     buildTransaction networkId nodeSocket changeAddress faucetUTxO [] [theOutput] >>= \case
       Left e -> throwIO $ FaucetFailedToBuildTx{reason = e}
       Right body -> do
-        submitTransaction networkId nodeSocket (sign faucetSk body)
+        let signedTx = sign faucetSk body
+        submitTransaction networkId nodeSocket signedTx
+        pure signedTx
```

Comment:
Keep this unchanged

```suggestion
        submitTransaction networkId nodeSocket (sign faucetSk body)
```

Reply:
but I need the function to return the `signedTx` for calling `awaitTransaction`.